
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>cap2.pipeline.short_read.humann2 &#8212; MetaSUB CAP 0.1.1 documentation</title>
    <link rel="stylesheet" href="../../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../../../" src="../../../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../../_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
   
  <link rel="stylesheet" href="../../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for cap2.pipeline.short_read.humann2</h1><div class="highlight"><pre>
<span></span>
<span class="kn">import</span> <span class="nn">luigi</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">from</span> <span class="nn">os.path</span> <span class="k">import</span> <span class="n">join</span><span class="p">,</span> <span class="n">dirname</span><span class="p">,</span> <span class="n">basename</span>

<span class="kn">from</span> <span class="nn">..utils.cap_task</span> <span class="k">import</span> <span class="n">CapTask</span>
<span class="kn">from</span> <span class="nn">..config</span> <span class="k">import</span> <span class="n">PipelineConfig</span>
<span class="kn">from</span> <span class="nn">..utils.conda</span> <span class="k">import</span> <span class="n">CondaPackage</span><span class="p">,</span> <span class="n">PyPiPackage</span>
<span class="kn">from</span> <span class="nn">..databases.uniref</span> <span class="k">import</span> <span class="n">Uniref90</span><span class="p">,</span> <span class="n">HumannIdTable</span>
<span class="kn">from</span> <span class="nn">..preprocessing.clean_reads</span> <span class="k">import</span> <span class="n">CleanReads</span>


<div class="viewcode-block" id="MicaUniref90"><a class="viewcode-back" href="../../../../modules_short_read.html#cap2.pipeline.short_read.MicaUniref90">[docs]</a><span class="k">class</span> <span class="nc">MicaUniref90</span><span class="p">(</span><span class="n">CapTask</span><span class="p">):</span>
    <span class="n">module_description</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    This module aligns reads to UniRef90 in</span>
<span class="s2">    preparation for functional profiling.</span>

<span class="s2">    Motivation: UniRef90 is a large database of functional genes. </span>

<span class="s2">    Negatives: Fairly slow. Many of the genes in UniRef90 are</span>
<span class="s2">    merely predicted functions.</span>

<span class="s2">    Note: this module currently uses Diamond not MiCA</span>
<span class="s2">    &quot;&quot;&quot;</span>
    <span class="n">MODULE_VERSION</span> <span class="o">=</span> <span class="s1">&#39;v0.2.0&#39;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pkg</span> <span class="o">=</span> <span class="n">CondaPackage</span><span class="p">(</span>
            <span class="n">package</span><span class="o">=</span><span class="s2">&quot;diamond&quot;</span><span class="p">,</span>
            <span class="n">executable</span><span class="o">=</span><span class="s2">&quot;diamond&quot;</span><span class="p">,</span>
            <span class="n">channel</span><span class="o">=</span><span class="s2">&quot;bioconda&quot;</span><span class="p">,</span>
            <span class="n">config_filename</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">config_filename</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">config</span> <span class="o">=</span> <span class="n">PipelineConfig</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">config_filename</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">out_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">out_dir</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">db</span> <span class="o">=</span> <span class="n">Uniref90</span><span class="p">(</span><span class="n">config_filename</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">config_filename</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reads</span> <span class="o">=</span> <span class="n">CleanReads</span><span class="o">.</span><span class="n">from_cap_task</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">tool_version</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">run_cmd</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{self.pkg.bin}</span><span class="s1"> --version&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">requires</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">pkg</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">reads</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">_module_name</span><span class="p">(</span><span class="bp">cls</span><span class="p">):</span>
        <span class="k">return</span> <span class="s1">&#39;diamond&#39;</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">dependencies</span><span class="p">(</span><span class="bp">cls</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">[</span><span class="s1">&#39;humann==3.0.0a4&#39;</span><span class="p">,</span> <span class="s1">&#39;diamond==0.9.32&#39;</span><span class="p">,</span> <span class="n">Uniref90</span><span class="p">,</span> <span class="n">CleanReads</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">output</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">{</span>
            <span class="s1">&#39;m8&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_target</span><span class="p">(</span><span class="s1">&#39;uniref90&#39;</span><span class="p">,</span> <span class="s1">&#39;m8.gz&#39;</span><span class="p">),</span>
        <span class="p">}</span>

    <span class="k">def</span> <span class="nf">_run</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">f</span><span class="s1">&#39;</span><span class="si">{self.pkg.bin}</span><span class="s1"> blastx &#39;</span>
            <span class="n">f</span><span class="s1">&#39;--threads </span><span class="si">{self.cores}</span><span class="s1"> &#39;</span>
            <span class="n">f</span><span class="s1">&#39;-d </span><span class="si">{self.db.diamond_index}</span><span class="s1"> &#39;</span>
            <span class="n">f</span><span class="s1">&#39;-q {self.reads.output()[&quot;clean_reads_1&quot;].path} &#39;</span>
            <span class="s1">&#39;--block-size 6 &#39;</span>
            <span class="n">f</span><span class="s1">&#39;| gzip &gt; {self.output()[&quot;m8&quot;].path} &#39;</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">run_cmd</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span></div>


<div class="viewcode-block" id="Humann2"><a class="viewcode-back" href="../../../../modules_short_read.html#cap2.pipeline.short_read.Humann2">[docs]</a><span class="k">class</span> <span class="nc">Humann2</span><span class="p">(</span><span class="n">CapTask</span><span class="p">):</span>
    <span class="n">module_description</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    This module provides functional profiles of microbiomes.</span>

<span class="s2">    Motivation: Functional profiles are sometimes more stable than</span>
<span class="s2">    taxonomic profiles and can provided metabolic and adaptive</span>
<span class="s2">    insights. </span>

<span class="s2">    Negatives: Functional profiling is somewhat less well benchmarked</span>
<span class="s2">    than taxonomic profiling and metabolic pathways are often based on</span>
<span class="s2">    model organisms.</span>

<span class="s2">    Notes: This class is named Humann2 for historical reasons. It uses</span>
<span class="s2">    humann3</span>
<span class="s2">    &quot;&quot;&quot;</span>
    <span class="n">MODULE_VERSION</span> <span class="o">=</span> <span class="s1">&#39;v0.3.0&#39;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">config</span> <span class="o">=</span> <span class="n">PipelineConfig</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">config_filename</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">humann</span> <span class="o">=</span> <span class="n">PyPiPackage</span><span class="p">(</span>
            <span class="n">package</span><span class="o">=</span><span class="s2">&quot;humann==3.0.0a4&quot;</span><span class="p">,</span>
            <span class="n">executable</span><span class="o">=</span><span class="s2">&quot;humann&quot;</span><span class="p">,</span>
            <span class="n">config_filename</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">config_filename</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">db</span> <span class="o">=</span> <span class="n">HumannIdTable</span><span class="o">.</span><span class="n">from_cap_task</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">out_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">out_dir</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alignment</span> <span class="o">=</span> <span class="n">MicaUniref90</span><span class="o">.</span><span class="n">from_cap_task</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">tool_version</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">run_cmd</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{self.pkg.bin}</span><span class="s1"> --version&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">requires</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">humann</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignment</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">_module_name</span><span class="p">(</span><span class="bp">cls</span><span class="p">):</span>
        <span class="k">return</span> <span class="s1">&#39;humann&#39;</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">dependencies</span><span class="p">(</span><span class="bp">cls</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">[</span><span class="s1">&#39;humann==3.0.0a4&#39;</span><span class="p">,</span> <span class="n">MicaUniref90</span><span class="p">,</span> <span class="n">HumannIdTable</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">output</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">{</span>
            <span class="s1">&#39;genes&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_target</span><span class="p">(</span><span class="s1">&#39;genes&#39;</span><span class="p">,</span> <span class="s1">&#39;tsv&#39;</span><span class="p">),</span>
            <span class="s1">&#39;path_abunds&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_target</span><span class="p">(</span><span class="s1">&#39;path_abunds&#39;</span><span class="p">,</span> <span class="s1">&#39;tsv&#39;</span><span class="p">),</span>
            <span class="s1">&#39;path_covs&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_target</span><span class="p">(</span><span class="s1">&#39;path_covs&#39;</span><span class="p">,</span> <span class="s1">&#39;tsv&#39;</span><span class="p">),</span>
        <span class="p">}</span>

    <span class="k">def</span> <span class="nf">_run</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">odir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span> <span class="o">+</span> <span class="s1">&#39;_humann2&#39;</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="n">odir</span> <span class="o">+</span> <span class="s1">&#39;/*genefamilies.tsv&#39;</span>
        <span class="n">abunds</span> <span class="o">=</span> <span class="n">odir</span> <span class="o">+</span> <span class="s1">&#39;/*pathabundance.tsv&#39;</span>
        <span class="n">covs</span> <span class="o">=</span> <span class="n">odir</span> <span class="o">+</span> <span class="s1">&#39;/*pathcoverage.tsv&#39;</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">f</span><span class="s1">&#39;</span><span class="si">{self.humann.bin}</span><span class="s1"> &#39;</span>
            <span class="n">f</span><span class="s1">&#39;--id-mapping </span><span class="si">{self.db.humann_id_table}</span><span class="s1"> &#39;</span>
            <span class="n">f</span><span class="s1">&#39;--input {self.alignment.output()[&quot;m8&quot;].path} &#39;</span>
            <span class="n">f</span><span class="s1">&#39;--output </span><span class="si">{self.sample_name}</span><span class="s1">_humann2 ; &#39;</span>
            <span class="s1">&#39;mv &#39;</span> <span class="o">+</span> <span class="n">genes</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">()[</span><span class="s1">&#39;genes&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">path</span> <span class="o">+</span> <span class="s1">&#39;; &#39;</span>
            <span class="s1">&#39;mv &#39;</span> <span class="o">+</span> <span class="n">abunds</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">()[</span><span class="s1">&#39;path_abunds&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">path</span> <span class="o">+</span> <span class="s1">&#39;; &#39;</span>
            <span class="s1">&#39;mv &#39;</span> <span class="o">+</span> <span class="n">covs</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">()[</span><span class="s1">&#39;path_covs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">path</span> <span class="o">+</span> <span class="s1">&#39;; &#39;</span>
            <span class="n">f</span><span class="s1">&#39;rm -r </span><span class="si">{self.sample_name}</span><span class="s1">_humann2;&#39;</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">run_cmd</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span></div>
</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../../index.html">MetaSUB CAP</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Documentation:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../architecture.html">Architecture and Design</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../modules.html">Pipeline modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../capalyzer.html">Creating and Parsing Data Tables</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../citation.html">Citing the CAP</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../license.html">License</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../../index.html">Documentation overview</a><ul>
  <li><a href="../../../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, David C. Danko.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.11</a>
      
    </div>

    

    
  </body>
</html>