
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Short Read &#8212; MetaSUB CAP 0.1.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Assembly" href="modules_assembly.html" />
    <link rel="prev" title="Preprocessing" href="modules_preprocessing.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="short-read">
<h1>Short Read<a class="headerlink" href="#short-read" title="Permalink to this headline">¶</a></h1>
<div class="figure" id="id1">
<a class="reference internal image-reference" href="_images/shortread.jpeg"><img alt="_images/shortread.jpeg" src="_images/shortread.jpeg" style="width: 600px;" /></a>
</div>
<p>The short read pipeline analyzes unassembled sequencing data to generate a number of useful scientific results including taxonomic profiles and functional profiles. This pipeline runs on data that has been preprocessed using the preprocessing pipeline.</p>
<div class="section" id="modules">
<h2>Modules<a class="headerlink" href="#modules" title="Permalink to this headline">¶</a></h2>
<div class="section" id="kraken2-taxonomic-profiling">
<h3>Kraken2 Taxonomic Profiling<a class="headerlink" href="#kraken2-taxonomic-profiling" title="Permalink to this headline">¶</a></h3>
<p>This module generates taxonomic profiles for each sample using <a class="reference external" href="https://ccb.jhu.edu/software/kraken2/">Kraken2</a>. Taxonomic profiles record how many reads in a sample can be confidently assigned to a set of microbial taxa. These profiles can be fed into downstream techniques to infer the relative abundance of different species in a sample or set of samples. For most microbiome studies this information underpins the rest of the results.</p>
<p>A number of different taxonomic profilers exist. Kraken2 is used because it is relatively resource efficient and <a class="reference external" href="https://lemmi.ezlab.org/#/">performs well on most accuracy benchmarks</a>. In the CAP Kraken2 is used with a large database containg all microbial genomes from RefSeq.</p>
<p>Kraken2 produces two files. A report file that summarizes the number of reads assigned to each taxa (and some diagnostic metrics) and a read assignment file which details what clade each read mapped to. An example of the output files from this module may be found on <a class="reference external" href="https://pangeabio.io/samples/eb8b8354-2cb3-43a7-8126-67c1c12cf684/analysis-results/3a250bb8-71cb-414c-b1d5-ee4c6cacbb95">Pangea</a>.</p>
<dl class="class">
<dt id="cap2.pipeline.short_read.Kraken2">
<em class="property">class </em><code class="descclassname">cap2.pipeline.short_read.</code><code class="descname">Kraken2</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cap2/pipeline/short_read/kraken2.html#Kraken2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cap2.pipeline.short_read.Kraken2" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="functional-profiling">
<h3>Functional Profiling<a class="headerlink" href="#functional-profiling" title="Permalink to this headline">¶</a></h3>
<p>This module identifies the abundance of microbial metabolic pathways using <a class="reference external" href="https://huttenhower.sph.harvard.edu/humann">HUMAnN</a>. These profiles are called functional profiles and are generally used for inferring what metabolites a microbiome can process and produce. The module works by first aligning reads to <a class="reference external" href="https://www.uniprot.org/help/uniref">UniRef90</a> using <a class="reference external" href="https://github.com/bbuchfink/diamond">Diamond</a> then by processing the resulting reads with HUMAnN.</p>
<p>Diamond produces one file as output: an M8 format blast tabular file. An example of the output file from Diamond may be found on <a class="reference external" href="https://pangeabio.io/samples/eb8b8354-2cb3-43a7-8126-67c1c12cf684/analysis-results/a709b48e-828d-44e7-879a-ac5989cb8f4a">Pangea</a>.</p>
<dl class="class">
<dt id="cap2.pipeline.short_read.MicaUniref90">
<em class="property">class </em><code class="descclassname">cap2.pipeline.short_read.</code><code class="descname">MicaUniref90</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cap2/pipeline/short_read/humann2.html#MicaUniref90"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cap2.pipeline.short_read.MicaUniref90" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="cap2.pipeline.short_read.Humann2">
<em class="property">class </em><code class="descclassname">cap2.pipeline.short_read.</code><code class="descname">Humann2</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cap2/pipeline/short_read/humann2.html#Humann2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cap2.pipeline.short_read.Humann2" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="mash-sketching">
<h3>MASH Sketching<a class="headerlink" href="#mash-sketching" title="Permalink to this headline">¶</a></h3>
<p><a class="reference external" href="https://mash.readthedocs.io/en/latest/index.html">MASH</a>. generates small sketches of sequencing data that can be used to quickly identify similar samples in an unbiased way. Mash sketches are based on finding a pre-set number of minimized kmers in a sample and finding the overlapping minimizers between two samples.</p>
<p>This module produces two output files: a small MASH sketch with 10,000 minimized k-mers and a large sketch  with 10,000,000 minimized k-mers. An example of the output files from this module may be found on <a class="reference external" href="https://pangeabio.io/samples/eb8b8354-2cb3-43a7-8126-67c1c12cf684/analysis-results/c98b1a8f-d568-4831-8c9a-072dd2561420">Pangea</a>.</p>
<dl class="class">
<dt id="cap2.pipeline.short_read.Mash">
<em class="property">class </em><code class="descclassname">cap2.pipeline.short_read.</code><code class="descname">Mash</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cap2/pipeline/short_read/mash.html#Mash"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cap2.pipeline.short_read.Mash" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="jellyfish-k-mer-counting">
<h3>Jellyfish K-mer Counting<a class="headerlink" href="#jellyfish-k-mer-counting" title="Permalink to this headline">¶</a></h3>
<p><a class="reference external" href="https://github.com/gmarcais/Jellyfish">Jellyfish</a>. counts the number of times each k-mer occurs in a sample to produce a k-mer profile. These profiles are useful to compare samples in an unbiased way and to search for particular sequences.</p>
<p>By default the CAP counts canonical 31-mers and 15-mers. All k-mers, including singletons, are counted. K-mer counting occurs after <span class="xref std std-ref">error correction</span> which substantially reduces the number of singletons compared to raw data.</p>
<p>Jellyfish produces two output files, both jellyfish archives (a custom format supported by jellyfish), one for 15-mers and one for 31-mers. An example of the output files from this module may be found on <a class="reference external" href="https://pangeabio.io/samples/eb8b8354-2cb3-43a7-8126-67c1c12cf684/analysis-results/0ba8eb21-cd93-4630-8be1-ce06cff39f09">Pangea</a>.</p>
<dl class="class">
<dt id="cap2.pipeline.short_read.Jellyfish">
<em class="property">class </em><code class="descclassname">cap2.pipeline.short_read.</code><code class="descname">Jellyfish</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cap2/pipeline/short_read/jellyfish.html#Jellyfish"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cap2.pipeline.short_read.Jellyfish" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">MetaSUB CAP</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Documentation:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="architecture.html">Architecture and Design</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">Pipeline modules</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="modules_qc.html">Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules_preprocessing.html">Preprocessing</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Short Read</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules_assembly.html">Assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules_databases.html">Databases</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="capalyzer.html">Creating and Parsing Data Tables</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="citation.html">Citing the CAP</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="modules.html">Modules</a><ul>
      <li>Previous: <a href="modules_preprocessing.html" title="previous chapter">Preprocessing</a></li>
      <li>Next: <a href="modules_assembly.html" title="next chapter">Assembly</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, David C. Danko.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.11</a>
      
      |
      <a href="_sources/modules_short_read.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>